zio-schema-json/shared/src/main/scala/zio/schema/codec/JsonCodec.scala

```diff
@@ -378,7 +378,9 @@ object JsonCodec {
 
     private def enumEncoder[Z](schema: Schema.Enum[Z], cfg: Config, cases: Schema.Case[Z, _]*): ZJsonEncoder[Z] =
       // if all cases are CaseClass0, encode as a String
-      if (schema.annotations.exists(_.isInstanceOf[simpleEnum])) {
+      if (schema.annotations.exists(_.isInstanceOf[simpleEnum]) && cases.forall(
```

Comment:
sine `@simpleEnum` is not added explicitly by the user, is this not an error in the macro?
And if the user would add it, the macro should fail.
But here, the user could have created a schema manually or added the annotation manually after the macro code.
So I think we should at least print a warning here. I would even prefer to fail.

Reply:
@pablf I think it is okay, if all leaves are objects or case classes without fields, that it is a simple enum. But why would you need the check in the json codec for that? The check there makes sense to prevent errors in case of an annotation where no annotation belongs, to raise an error. But if leaves are simple cases and the json codec still can't handle it, it is a different error and this would not be the fix.
I think the goals should be
- simple leaves = simple enum
- simple leaves with intermediate traits should be handled by codecs as if the intermediate trait is missing
- no simple leaves but annotated => macro fails
- no simple leaves but annotated => codec fails

Reply:
Maybe we need more issues

Reply:
I think it should not. At least I do not see any value in it. @jdegoes wdyt?

Reply:
@987Nabil Why not derive intermediate traits??

Reply:
I don't say it's not good. I say, is there a use case for it? Is there value in it? Or does it just make the use case of simple enum harder without a clear benefit.

Anyhow, if we have intermediate traits derived, it should be the same for Scala 2 and 3 and leave traits (no subtypes) should be ignored for the question of "Is this a simple enum".

Reply:
It seems like not deriving intermediates makes the system "lossy", which is something I like to avoid on principle even if I don't have a compelling use case for it.

Reply:
I am fine with that. Then the Scala 2 macro should do the same. And the structure should be interpreted correctly regarding simple enum